storage/txcache/eviction.go

```diff
@@ -123,18 +138,30 @@ func (cache *TxCache) evictHighNonceTransactions() (uint32, uint32) {
 		}
 	})
 
+	if len(txsToEvict) == 0 {
+		return 0, 0
+	}
+
 	return cache.doEvictItems(txsToEvict, sendersToEvict)
 }
 
-// evictSendersWhileTooManyTxs removes transactions
-// Eviction happens in ((transaction count) - CountThreshold) / NumOldestSendersToEvict + 1 steps
-// One batch of senders is removed in each step
 func (cache *TxCache) evictSendersWhileTooManyTxs() (step uint32, countTxs uint32, countSenders uint32) {
-	batchesSource := cache.txListBySender.GetListsSortedByOrderNumber()
-	batchSize := cache.evictionConfig.NumOldestSendersToEvict
+	return cache.evictSendersWhile(cache.areThereTooManyTxs, SortByOrderNumberAsc)
+}
+
+func (cache *TxCache) evictSendersWhileTooManyBytes() (step uint32, countTxs uint32, countSenders uint32) {
+	return cache.evictSendersWhile(cache.areThereTooManyBytes, SortBySmartScoreAsc)
+}
+
+// evictSendersWhileTooManyTxs removes transactions in a loop, as long as "stopCondition" is true
+// One batch of senders is removed in each step
+// Before starting the loop, the senders are sorted as specified by "sendersSortKind"
+func (cache *TxCache) evictSendersWhile(stopCondition func() bool, sendersSortKind txListBySenderSortKind) (step uint32, countTxs uint32, countSenders uint32) {
+	batchesSource := cache.txListBySender.GetListsSortedBy(sendersSortKind)
```

Comment:
as I see, every time the eviction conditions are met (too many senders, too many txs, too many bytes) we sort the entire set of senders by some criteria, some simpler like the order number, some more complex like a computed cost.

This will be very cpu intensive, especially when we have many senders, and if we have to run this with a high frequency I think it becomes a bottleneck, since previous sorts and computations of costs from previous eviction conditions are not "sticky" and don't improve the time for next computations.

Can you think of a possible way to have the order precalculated for the eviction and not have to do any sorting and any computation, just iteration and removal?

Reply:
Removed sorting overall. Now we sort the senders at insertion time as well.